keps/sig-network/NNNN-kubelet-probes/README.md

```diff
@@ -0,0 +1,1039 @@
+  network, so that (among other things) kubelet does not need to have
+  any particular pod-network security privileges.
+
+    - (For pods that really have a strong use case for checking
+      connectivity in some non-standard way, such as via a load
+      balancer IP, it would still be possible for a user to write an
+      exec probe that uses `curl` or the like to connect to that IP
+      from the pod itself.)
+
+    - If we implement this by creating a new set of probe types,
+      rather than by modifying the semantics of the existing probes,
+      then it should be possible to block the old probe types via Pod
+      Security admission, so that administrators can create
+      AdminNetworkPolicies that don't include the kubelet probe hole.
+
+- Allow dual-stack / wrong-single-stack probing, either based on
+  information in the probe definition (e.g. `ipFamily: IPv6` or
+  `ipFamilyPolicy: RequireDualStack`) or by always doing "[Happy
+  Eyeballs]" (probing IPv4 and IPv6 in parallel and only requiring one
+  to succeed).
+
+- Make it easy for users who aren't doing "weird" things (eg,
+  readiness probe via LoadBalancer) to transition from the old probe
+  system to the new one.
+
+    - This might mean that they don't need to do anything, or it might
+      mean that there's an easy transition, like just changing
+      `tcpSocket:` to `tcp:` in their pod's probe definition.
+
+[`TCPSocketAction.Host`]: https://github.com/kubernetes/kubernetes/blob/v1.28.0/staging/src/k8s.io/api/core/v1/types.go#L2248
+[`HTTPGetAction.Host`]: https://github.com/kubernetes/kubernetes/blob/v1.28.0/staging/src/k8s.io/api/core/v1/types.go#L2219
+[Pod Security admission]: https://kubernetes.io/docs/concepts/security/pod-security-admission/
+[Happy Eyeballs]: https://tools.ietf.org/html/rfc8305
+
+### Non-Goals
+
+- Adding entirely-new probe types.
+
+- Changing the definition of NetworkPolicy v1 to remove the "kubelet
+  can access all local pods" hole, even in the case where the cluster
+  is configured in a way to forbid the use of probes that require the
+  hole. Even if _kubelet_ no longer needs to access pods, some users
+  likely depend on that rule to allow access to some pods from other
+  host-level processes, and there is no other way in NetworkPolicy v1
+  to write such a rule if we remove the built-in one. (However,
+  AdminNetworkPolicy and any future successor to NetworkPolicy should
+  be designed to not have this rule by default.)
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### API Changes to Fix the SSRF
+
+Regardless of what else we do, we will deprecate the
```

Comment:
can we really deprecate them? I know users that depend on this functionality

Reply:
Do you have examples?

If we know what people are using it for, we can try to come up with an API that allows the "sane" use cases without allowing it to be used for an attack.

Reply:
https://docs.tilt.dev/api.html#api.http_get_action
https://github.com/tilt-dev/tilt/blob/3359b56077d1bff23de1d179cb99431eefdd3c35/internal/tiltfile/api/__init__.py#L1385

https://stackoverflow.com/questions/63082309/liveness-probe-of-one-pod-via-another
https://stackoverflow.com/questions/75811572/executing-readiness-httpget-readiness-probe-failed-no-such-host

I remember some user in kubernetes/kubernetes explaining they used the probe to query an external http service but I can not find it ...

Reply:
> https://docs.tilt.dev/api.html#api.http_get_action https://github.com/tilt-dev/tilt/blob/3359b56077d1bff23de1d179cb99431eefdd3c35/internal/tiltfile/api/__init__.py#L1385

That just shows that they're re-exporting the (full) API, not that anyone is actually using it.

> https://stackoverflow.com/questions/63082309/liveness-probe-of-one-pod-via-another

"pod B depends on pod A working so I want pod B's liveness probe to check pod A (via a service externalIP)".

That's a little goofy... Clearly it would be architecturally nicer (if more complicated) to make pod B implement a proper health check itself.

Also, this health check, as written, depends on the existence of the NetworkPolicy hole (since it eventually results in a kubelet-to-pod-A connection) so even if we allowed it, it wouldn't work with AdminNetworkPolicy...

> https://stackoverflow.com/questions/75811572/executing-readiness-httpget-readiness-probe-failed-no-such-host

That's a little more plausible since they're using a service IP that resolves to the pod they're probing, rather than an entirely unrelated pod. This still requires the NetworkPolicy hole though, and this goes back to the "[are probes about server readiness or network connectivity?](https://github.com/kubernetes/enhancements/pull/4558/files#diff-7e38dd798d6b784375dfe279505947c18ea870f56da9fe4d87395c96df27e13dR190-R199)" question

> I remember some user in kubernetes/kubernetes explaining they used the probe to query an external http service but I can not find it ...

That seems even more dubious than the someone-else's-externalIP example.

At any rate, all of these _could_ be implemented via exec probes if it was no longer possible to do them via http probes.

Reply:
just to be clear, I'm in favor of this, I also opened the issue some time ago to move the TCP and HTTP probes inside the Pods, and I think that they are different ways of doing this keeping the functionality ... is just that realistically speaking the API and the semantics are v1 and changing them will mean that is a backwards incompatible change :(

Reply:
For this to work, k8s will need knowledge of the network namespace. It sounds like we want to execute into the netns and run the health checks?

Reply:
https://github.com/kubernetes/kubernetes/issues/102613

and there are other tech details in the doc suggested, as reusing existing CRI exec rpc

Reply:
> For this to work, k8s will need knowledge of the network namespace. It sounds like we want to execute into the netns and run the health checks?

There are multiple ideas suggested. And as noted in a comment below, although this KEP was just filed this week, I actually wrote all of it several months ago, so the existing text doesn't consider KNI at all.